parsers/BD.py

```diff
@@ -1,299 +1,295 @@
 #!/usr/bin/env python3
 
-"""Parser for Bangladesh."""
-
 from datetime import datetime
 from logging import Logger, getLogger
-from typing import List, Optional
+from typing import Any, Dict, List, Optional, Union
 
 import arrow
-import pandas as pd
-from requests import Session
-
-from .lib.exceptions import ParserException
-
-GENERATION_MAPPING = {
-    "Gas (Public)": "gas_public",
-    "Gas (Private)": "gas_private",
-    "Oil (Public)": "oil_public",
-    "Oil (Private)": "oil_private",
-    "Coal": "coal",
-    "Hydro": "hydro",
-    "Solar": "solar",
-}
-
-
-OLD_GENERATION_MAPPING = {
-    "GAS": "gas_public",
-    "P.GEN(GAS).": "gas_private",
-    "OIL": "oil_public",
-    "P.GEN(OIL).": "oil_private",
-    "COAL": "coal",
-    "HYDRO": "hydro",
-    "SOLAR": "solar",
-}
-
-
-def timestamp_converter(raw_timestamp, target_datetime):
-    """Converts string timestamp (e.g. 10:00:00) to arrow object."""
-
-    hour, minute = raw_timestamp.split(":")[0:2]
+import pytz
+from bs4 import BeautifulSoup, Tag
+from requests import Response, Session
+
+from parsers.lib.exceptions import ParserException
+
+# Power Grid Company of Bangladesh: erp.pgcb.gov.bd
+# Has table (also historical) of production, consumption and exchange.
+# This table gets updated batch-wise every few hours, so most of the time, the delay will be >2h.
+
+tz = pytz.timezone("Asia/Dhaka")
+latest_url = "https://erp.pgcb.gov.bd/web/generations/view_generations"
+url_by_date = (
+    "https://erp.pgcb.gov.bd/web/generations/view_generations?search="  # DD-MM-YYYY
+)
+
+TABLE_HEADERS = [
+    "Date",
+    "Time",
+    "Generation(MW)",
+    "Demand(MW)",
+    "Loadshed",
+    "Gas",
+    "Liquid Fuel",
+    "Coal",
+    "Hydro",
+    "Solar",
+    "Bheramara HVDC",
+    "Tripura",
+    "Remarks",
+]
+
+
+def table_entry_to_float(entry: str):
+    """
+    Helper function to handle empty cells in table.
+    """
+    if entry == "":
+        return 0.0
```

Comment:
If the entry is missing ("") then we should return None instead of 0. 0 should never be set manually (there are rare cases where negative production is reported and we set it to 0 but this is not the case here).

Reply:
Done at b688086. Also checking for None values when actually putting the data into the production/consumption/exchange dicts.